File: test/SaltarelleProjectTests/CoreSubset/CoreSlim/Logging/ConsoleLogAppender.cs
// ---------------------------------------------------------------------------------------------------------------------
// <copyright file="ConsoleLogAppender.cs" company="Tableau Software">
//   This file is the copyrighted property of Tableau Software and is protected by registered patents and other
//   applicable U.S. and international laws and regulations.
//
//   Unlicensed use of the contents of this file is prohibited. Please refer to the NOTICES.txt file for further details.
// </copyright>
// ---------------------------------------------------------------------------------------------------------------------

namespace Tableau.JavaScript.Vql.Core
{
    using System;
    using System.Collections.Generic;
    using System.Html;

    /// <summary>
    /// An appender that writes to console.log.
    /// </summary>
    public class ConsoleLogAppender : BaseLogAppender
    {
        //// ===========================================================================================================
        //// Member Variables
        //// ===========================================================================================================

        public static readonly LogAppenderInstance<ConsoleLogAppender> GlobalAppender = new LogAppenderInstance<ConsoleLogAppender>(() => new ConsoleLogAppender());

        private JsDictionary<string, object> levelMethods;

        //// ===========================================================================================================
        //// Constructors
        //// ===========================================================================================================

#if DEBUG

        static ConsoleLogAppender()
        {
            // by default we're going to enable console.log for Info+ in debug
            GlobalAppender.EnableLogging((_, loggerLevel) => loggerLevel >= LoggerLevel.Info);
        }

#endif

        private ConsoleLogAppender()
        {
        }

        //// ===========================================================================================================
        //// Methods
        //// ===========================================================================================================
        protected override void LogInternal(Logger source, LoggerLevel level, string message, object[] args)
        {
            if (Script.TypeOf(((dynamic)typeof(Window)).console) != "object")
            {
                return;
            }

            message = source.Name + ": " + message;

            JsArray<object> consoleArgs = new JsArray<object>();
            consoleArgs = consoleArgs.Concat(message).Concat(args);

            try
            {
                ((dynamic)typeof(Function).Prototype).apply.call(this.GetConsoleMethod(level), ((dynamic)typeof(Window)).console, consoleArgs);
            }
            catch
            {
                // ignore
            }
        }

        private object GetConsoleMethod(LoggerLevel level)
        {
            dynamic console = TypeUtil.GetField<object>(Window.Self, "console");
            if (this.levelMethods == null)
            {
                this.levelMethods = new JsDictionary<string, object>();
                this.levelMethods[LoggerLevel.Debug.ToString()] = console.log;
                this.levelMethods[LoggerLevel.Error.ToString()] = console.error;
                this.levelMethods[LoggerLevel.Info.ToString()] = console.info;
                this.levelMethods[LoggerLevel.Warn.ToString()] = console.warn;
            }

            return this.levelMethods[level.ToString()] ?? console.log;
        }
    }
}

File: E2ETestResults/CoreSubset/CoreSlim/Logging/ConsoleLogAppender.ts
import { BaseLogAppender } from './BaseLogAppender';

import { LogAppenderInstance } from './LogAppenderInstance';

import { Logger, LoggerLevel } from './Logger';

import 'mscorlib';

import { TypeUtil } from 'NativeJsTypeDefs';

/**
 * An appender that writes to console.log.
 */
export class ConsoleLogAppender extends BaseLogAppender {
  public static readonly globalAppender: LogAppenderInstance<ConsoleLogAppender> = new LogAppenderInstance<ConsoleLogAppender>(() => new ConsoleLogAppender());

  private levelMethods: { [key: string]: any };

  // Converted from the C# static constructor - it would be good to convert this
  // block to inline initializations.
  public static __ctor() {
    ConsoleLogAppender.globalAppender.enableLogging((_, loggerLevel) => loggerLevel >= LoggerLevel.Info);
  }

  private constructor() { }

  protected logInternal(source: Logger, level: LoggerLevel, message: string, args: any[]): void {
    if (typeof (<any>window).console !== 'object') {
      return;
    }
    message = source.name + ': ' + message;
    let consoleArgs: any[] = [];
    consoleArgs = consoleArgs.concat([message]).concat([args]);
    try {
      (<any>Function.prototype).apply.call(this.getConsoleMethod(level), (<any>window).console, consoleArgs);
    } catch { }
  }

  private getConsoleMethod(level: LoggerLevel): any {
    let console: any = window.self['console'];
    if (this.levelMethods === null) {
      this.levelMethods = {};
      this.levelMethods[LoggerLevel.Debug.toString()] = console.log;
      this.levelMethods[LoggerLevel.Error.toString()] = console.error;
      this.levelMethods[LoggerLevel.Info.toString()] = console.info;
      this.levelMethods[LoggerLevel.Warn.toString()] = console.warn;
    }
    return this.levelMethods[level.toString()] || console.log;
  }
}

// Call the static constructor
ConsoleLogAppender.__ctor();
